rocprim/include/rocprim/device/detail/device_reduce_by_key.hpp

```diff
@@ -44,6 +44,16 @@ namespace detail
 template<class Key, class Value>
 struct carry_out
 {
+    ROCPRIM_HOST_DEVICE inline
+    carry_out& operator=(carry_out rhs)
```

Comment:
Why is that necessary?

Reply:
Seems that without this copy operator, the compiler will complain about using a default copy operator which has attribute `__host__` but is being called by `__device__` function. so I added this to add ROCPRIM_HOST_DEVICE.

Reply:
Shouldn't it be generated by the compiler?

Reply:
I think it's might be a CUDA spec limitation where `__device__` functions can only call explicitly defined `__device__` functions. I will discuss it with my colleague

Reply:
It seems that the spec suggests without the attributes, it should be compiled for host code only:
```
The __host__ qualifier declares a function that is:
 Executed on the host,
 Callable from the host only.
It is equivalent to declare a function with only the __host__ qualifier or to declare
it without any of the __host__, __device__, or __global__ qualifier; in either
case the function is compiled for the host only. 
```

Reply:
Hmm, ok, I guess I worked with `hcc` too much and assumed implicitly generated special member functions like copy assignment operator get required attributes as needed.

Anyway, instead of implementing them it's better to write:

```cpp
ROCPRIM_HOST_DEVICE inline
scan_by_key_pair& operator=(const scan_by_key_pair& rhs) = default;
```
That way there's no need for modification if member variables are changed.

Reply:
I tried to do that at first, but it seems that the `default` is only defined with `__host__` :( 
```
In file included from /root/rocPRIM/test/rocprim/test_hip_device_run_length_encode.cpp:36:
In file included from /root/rocPRIM/rocprim/include/rocprim/rocprim.hpp:74:
In file included from /root/rocPRIM/rocprim/include/rocprim/device/device_reduce_by_key_hip.hpp:34:
/root/rocPRIM/rocprim/include/rocprim/device/detail/device_reduce_by_key.hpp:67:70: error: reference to
      __host__ function 'operator=' in __host__ __device__ function
    scan_by_key_pair& operator=(const scan_by_key_pair& rhs) = default;
                                                                     ^
```

Reply:
Can you try defaulting it separately for host and device? or better: only for device?

Reply:
Yes that works better, thanks Jakub